microsoft/testsuites/security/fips.py

```diff
@@ -18,10 +27,106 @@
     Tests the functionality of FIPS enable
     """,
 )
-class Fips(TestSuite):
-    _expected_failure_pattern = re.compile(
-        "Error setting digest\r\n.*EVP_DigestInit_ex:disabled for FIPS.*", re.M
+class FipsTests(TestSuite):
+    @TestCaseMetadata(
+        description="""
+            Ensures that an AZL machine is fips enabled.
+        """,
+        priority=1,
+        requirement=simple_requirement(
+            supported_os=[CBLMariner],
+        ),
+    )
+    def verify_azl_fips_is_enabled(
+        self, log: Logger, node: Node, variables: Dict[str, Any]
+    ) -> None:
+        # Skip the test if the image is not FIPS enabled.
+        self._ensure_fips_expectations(log, node, variables, should_be_fips=True)
+
+        # Ensure the system is FIPS enabled.
+        node.tools[Fips].assert_fips_mode(True)
+
+        log.info("FIPS is enabled and working correctly.")
+
+    @TestCaseMetadata(
+        description="""
+            Ensures that an AZL machine is not FIPS enabled.
+        """,
+        priority=3,
+        requirement=simple_requirement(
+            supported_os=[CBLMariner],
+        ),
+    )
+    def verify_azl_fips_is_disabled(
+        self, log: Logger, node: Node, variables: Dict[str, Any]
+    ) -> None:
+        # Skip the test if the image is FIPS enabled.
+        self._ensure_fips_expectations(log, node, variables, should_be_fips=False)
+
+        # Ensure the system is not FIPS enabled.
+        node.tools[Fips].assert_fips_mode(False)
+
+        log.info("FIPS is disabled and properly.")
+
+    @TestCaseMetadata(
+        description="""
+            This test case will
+            1. Enable FIPS on the AZL machine
+            2. Restart the machine
+            3. Verify that FIPS was enabled properly
+        """,
+        priority=2,
+        requirement=simple_requirement(
+            supported_os=[CBLMariner],
+        ),
+    )
+    def verify_azl_fips_enable(
+        self, log: Logger, node: Node, variables: Dict[str, Any]
+    ) -> None:
+        # Skip the test if the image is already FIPS enabled.
+        self._ensure_fips_expectations(log, node, variables, should_be_fips=False)
+
+        # Enable FIPS on the system and make sure it is worked.
+        azl_fips = node.tools[Fips]
+        azl_fips.enable_fips()
+        node.reboot()
+        azl_fips.assert_fips_mode(True)
+
+        log.info("Successfully enabled FIPS.")
+
+        # Re-disable FIPS to make sure the test can be run multiple times.
+        azl_fips.disable_fips()
+        node.reboot()
```

Comment:
Recover the fips status, because we don't know the order of test cases. If the test case run before verifying initial state case, the initial state case will be failed. If it's hard to guarantee reset, you can mark the environment is dirty at end of this test case. So this env won't be reused.